Dask merge concat fill value

bjlittle opened the pull request:
This PR refines the approach taken with handling the `cube.fill_value` by merge and concatenate.

This PR extends the expected behaviour of `numpy` with regards to how it manages the `fill_value` when combining masked arrays and/or ndarrays to cubes (it is valid for a cube to have a `fill_value` even though it has a ndarray payload).

In summary, when combining cubes via merge or concatenate, the resultant cube will have:
* a `fill_value` that is non-`None` **iff** all cubes have the same non-`None` `fill_value`
* a `fill_value` that is `None`, if:
   * at least one cube has a `fill_value` of `None`, or
   * the candidate cube and the proto-cube have a different `fill_value`

This simply means that, 
* if the `cube.fill_value` is set to a non-`None` value, and the cube has a masked payload, then the masked array returned by `cube.data` will have a `fill_value` that matches the `cube.fill_value`.
* if the `cube.fill_value` is set to `None`, and the cube has a masked payload, then the masked array returned by `cube.data` will have the default `fill_value` generated by `numpy` for that specific `dtype`.

bjlittle commented:
Ping @djkirkham and @lbdreyer ... :smile:

lbdreyer commented on lib/iris/_merge.py:
```diff
@@ -407,11 +407,6 @@ def match(self, other, error_on_mismatch):
-            # Allow fill-value promotion if either fill-value is None.
-            if self.fill_value is not None and other.fill_value is not None:
-                msg = 'cube fill value differs: {} != {}'
-                msgs.append(msg.format(self.fill_value, other.fill_value))
```
Why did you choose to allow for merging and concatenating cubes with different fill values?
I realise this is not a simple question so I can call you to discuss this if that's easier?

bjlittle replied:
@lbdreyer This is more in alignment with the behaviour of `upstream/master`, which doesn't discriminate based on `fill_value`. So no breaking or surprising behaviour for the end user.

It also aligns with the behaviour of `numpy`, which will merge/concatenate arrays with different `fill_values`, but the resultant array has a `fill_value` that is replaced with the `numpy` nominated default for that `dtype`. 

With regards to cubes, we're doing the same - if we have two cubes that only differ on `fill_value`, say one is `1234` and the other is `4321`, then they will merge/concatenate into one cube, but the resultant cube will have a `fill_value` of `None. When the user does a `cube.data` to get the data, the resulting masked array (assuming the payload realizes is masked) will have a `fill_value` that is the `numpy` nominated default.

lbdreyer commented on lib/iris/tests/test_concatenate.py:
```diff
@@ -748,6 +730,49 @@ def test_concat_2x2d_aux_xy_bounds(self):
+            np_fill_value = ma.masked_array(0, dtype=result.dtype).fill_value
+            self.assertEqual(result.data.fill_value, np_fill_value)
+
+    def test_fill_value_invariant_to_order__different_non_None(self):
```
Other than testing the different order, is this much different to [test_concat_masked_2y2d__default_fill_value_from_diff](https://github.com/SciTools/iris/pull/2520/files#diff-af298a2170476a32abb0c94b312b4ff5R438)?

I don't understand what you gain from doing all the permutations?

lbdreyer replied:
Okay I understand that this concatenates cubes with x's like
```
concat [0, 1] [2, 3] [4, 5] => [0, 1, 2, 3, 4, 5]
concat [0, 1] [4, 5] [2, 3] => [0, 1, 2, 3, 4, 5]
concat [4, 5] [0, 1] 2, 3] => [0, 1, 2, 3, 4, 5]
```

and checking the fill value is correct after that

lbdreyer commented on lib/iris/tests/test_concatenate.py:
```diff
@@ -748,6 +730,49 @@ def test_concat_2x2d_aux_xy_bounds(self):
         self.assertEqual(len(result), 1)
         self.assertEqual(result[0].shape, (2, 4))
 
+    def test_fill_value_invariant_to_order__same_non_None(self):
```
This test naming was a bit confusing to me. I only understood what `__same_non_None` meant when I started writing a comment asking you what it is. I am going to try to think of a different name

bjlittle replied:
Yeah, test naming ... ugh ... you are right, I've don my thinking cap :cowboy_hat_face:

lbdreyer commented on lib/iris/tests/test_concatenate.py:
```diff
@@ -335,113 +336,93 @@ def test_order_difference(self):
-            cubes.concatenate_cube()
-
 
 class Test2D(tests.IrisTest):
```
You never test the fill value when all inputs that get concatenated are unmasked. You could always add a quick check to the test on [L535](https://github.com/bjlittle/iris/blob/fb0a3a51b308d7966c408b6ec2e498667cff29c9/lib/iris/tests/test_concatenate.py#L535)

lbdreyer commented on lib/iris/tests/test_merge.py:
```diff
@@ -145,81 +145,143 @@ def _make_cube(self, data, dtype=np.dtype('int32'), fill_value=None,
+        return result
+
+    def _check_fill_value(self, result, fill0, fill1):
+        fill_value = self._expected_fill(fill0, fill1)
```
Rename to be `expected_fill_value` or `expected_cube_fill_value`?
May make it easier to follow

lbdreyer commented on lib/iris/tests/test_merge.py:
```diff
@@ -145,81 +145,143 @@ def _make_cube(self, data, dtype=np.dtype('int32'), fill_value=None,
+                                        [False, True])
+        fill_combos = itertools.product([None, fill_value],
+                                        [fill_value, None])
+        self.combos = itertools.product(lazy_combos, fill_combos)
```
Was it too complicated to have another `itertools.product` for ndarray and masked?

bjlittle replied:
I think I'm a wee bit wary of folding that in, as it seems like we'll end up with a single test that does absolutely everything and will be a tad impenetrable ... is that okay?

lbdreyer replied:
Sure! makes sense

lbdreyer commented:
This looks very close!

The testing of concatenate isn't less complete than the testing of merge, but there are endless combinations (masked/unmasked, fill_value_same/fill_value_diff/fill_value_None, lazy/real etc) that we can test for... I want to go through some possible *important* combinations that we are missing.
but otherwise I'm happy with the changes.

lbdreyer commented:
I've looked into testing of concatenate a little more and ignoring the endless combinations we could test I think there are two missing combinations that we should at least consider.

1) There doesn't appear to be a test for concatenating cubes that are masked and unmasked where both have a fill value of None.

2) Also there isn't testing of fil value when both cubes are unmasked.

bjlittle commented:
@lbdreyer Okay, so I've addressed your review actions in the simplest way possible. Adding a concatenate equivalent to the merge `TestDataMergeCombos` isn't really practical and would involve quite a bit of effort, so I've added the missing test coverage through new tests and adding to existing tests.

djkirkham commented on lib/iris/_concatenate.py:
```diff
@@ -727,10 +718,14 @@ def register(self, cube, axis=None, error_on_mismatch=False,
+                if cube_signature.fill_value is None or \
+                        cube_signature.fill_value != fill_value:
+                    # Demote the fill value to the default.
+                    self._cube_signature.fill_value = None
```
This can be simplified to
```
if cube_signature.fill_value != self._cube_signature.fill_value:
    self._cube_signature.fill_value = None
```

bjlittle replied:
Yup, also applies to `_merge.py` :+1:

djkirkham commented on lib/iris/_merge.py:
```diff
@@ -1285,15 +1280,19 @@ def register(self, cube, error_on_mismatch=False):
+                if other.fill_value is None or \
+                        cube_signature.fill_value != other.fill_value:
+                    # Demote the fill value to the default.
+                    signature = self._build_signature(self._source,
```
Why do we need a call to `_build_signature` here, but in the concatenate case you can just set `fill_value` to None?

bjlittle replied:
One is a class the other is a specialization of a `namedtuple`

djkirkham commented on lib/iris/tests/test_concatenate.py:
```diff
@@ -335,113 +337,117 @@ def test_order_difference(self):
-        self.assertEqual(result[0].data.fill_value, fill_value)
+        self.assertIsNone(result[0].fill_value)
+        np_fill_value = ma.masked_array(0, dtype=result[0].dtype).fill_value
+        self.assertEqual(result[0].data.fill_value, np_fill_value)
```
I'm wondering if we need this check. I think perhaps it should be part of the contract that if the input fill values don't match then the fill value you get out on the array shouldn't be relied upon to have any particular value. That said, we are making an effort to make sure the correct fill value is placed on the data array when it _is_ valid, so I'm not sure.

djkirkham replied:
I'd be happy either way on both points to be honest

djkirkham commented on lib/iris/tests/test_concatenate.py:
```diff
@@ -748,6 +760,49 @@ def test_concat_2x2d_aux_xy_bounds(self):
+            self.assertIsNone(result.fill_value)
+            np_fill_value = ma.masked_array(0, dtype=result.dtype).fill_value
+            self.assertEqual(result.data.fill_value, np_fill_value)
+
```
Nice test coverage!

bjlittle replied:
I try :wink: 

At least we're trying to join the dots on what's happening here ... so if it changes, a whole bunch of tests are now gonna complain .... which is a good thing!

lbdreyer commented on lib/iris/tests/test_merge.py:
```diff
@@ -146,14 +146,14 @@ def _make_cube(self, data, dtype=np.dtype('int32'), fill_value=None,
 
     def _check_fill_value(self, result, fill0, fill1):
-        fill_value = self._expected_fill(fill0, fill1)
+        fill_value = self._expected_fill_value(fill0, fill1)
```
When I said I would prefer a rename I meant the variable, not the method, 
i.e. ```expected_fill_value = self._expected_fill(fill0, fill1)```
and then the self.assertEqual checks would be:

```self.assertEqual(data.fill_value, expected_fill_value)```

lbdreyer commented on lib/iris/tests/test_merge.py:
```diff
@@ -145,81 +145,143 @@ def _make_cube(self, data, dtype=np.dtype('int32'), fill_value=None,
+
+    def _check_fill_value(self, result, fill0, fill1):
+        fill_value = self._expected_fill_value(fill0, fill1)
+        if fill_value is None:
```
When I said I would prefer a rename I meant the variable, not the method, 
i.e. ```expected_fill_value = self._expected_fill(fill0, fill1)```
and then the self.assertEqual checks would be:

```self.assertEqual(data.fill_value, expected_fill_value)```

lbdreyer commented:
@bjlittle if you adress my [last comment](https://github.com/SciTools/iris/pull/2520/files/d44458d1125f3a2f95092e46a511c2b7a32ad05c#r114774901) I will happily merge this in

bjlittle commented:
@lbdreyer Thanks :+1:
